doc/decisions/functions_with_buffers.md

```diff
@@ -0,0 +1,43 @@
+# Functions copying into buffers
+
+## Problem
+
+Currently the way functions like keyGetName() work is by passing a buffer with
+a maxSize and if the buffer is large enough, the value gets copied into the
+buffer. This leads to the user having to write a lot of surrounding boilerplate
+code, checking for the size of every value / name they want to copy into a buffer.
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+## Decision
+
+- Remove Functions:
+  - keyGetName()
+  - keyGetUnescapedName()
+  - keyGetBaseName()
+  - keyGetString()
+  - keyGetBinary()
+- add documentation that you should use strncpy() / memcpy() instead:
```

Comment:
strndup seems to be the easiest way for users who simply want a duplicate string

```suggestion
- add documentation that you should use strndup() instead:
```

the second argument should be a call to keyGet*Size

Reply:
> strndup seems to be the easiest way for users who simply want a duplicate string

True, but `strndup(keyName(k), size)` is not a replacement for what `keyGetName(k, buffer, size)` does, because it cannot copy into a pre-existing buffer.

Reply:
Good point, we need to be clear who we want to address with our API docu. I was thinking about users who want to use a name or value beyond the life-time of the key object (and give them the easiest way to do so).

For copying into other buffers there are endless ways, that is why I thought it is better we do not describe this. @kodebach which kind of user (or use case) were you thinking of? Did you think about the migration guide for users of our previous API? Shouldn't that be better in the release notes?

Reply:
> Shouldn't that be better in the release notes?

I was thinking about release notes, or similar documentation for how e.g. `keyGetName` should be replaced. I assumed this is what the line is about because it says "instead", which implies "to replace".

> use a name or value beyond the life-time of the key object

there is also endless ways to do that ;)

I don't think we need to mention how to do this in the Doxygen of e.g. `keyName()`. Saying `keyName(k)` returns an internal pointer that will only be valid until you call `keyDel (k)` or any function that modifies the name of `k`, should be enough. Elektra's docs are not the right place to explain "How to copy a string C".

Reply:
I agree!

```suggestion
- add documentation in API docu about life-time and add in release notes that you should use strncpy() / memcpy() instead:
```